controller-server/src/main/java/com/yahoo/vespa/hosted/controller/notify/Notifier.java

```diff
@@ -0,0 +1,68 @@
+package com.yahoo.vespa.hosted.controller.notify;
+
+import com.yahoo.vespa.hosted.controller.api.integration.organization.Mail;
+import com.yahoo.vespa.hosted.controller.api.integration.organization.Mailer;
+import com.yahoo.vespa.hosted.controller.notification.Notification;
+import com.yahoo.vespa.hosted.controller.persistence.CuratorDb;
+import com.yahoo.vespa.hosted.controller.tenant.CloudTenant;
+import com.yahoo.vespa.hosted.controller.tenant.TenantContacts;
+
+import java.util.Collection;
+import java.util.Formatter;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * Notifier is responsible for dispatching user notifications to their chosen Contact points.
+ *
+ * @author enygaard
+ */
+public class Notifier {
+    private final CuratorDb curatorDb;
+    private final Mailer mailer;
+
+    private static final Logger log = Logger.getLogger(Notifier.class.getName());
+
+    public Notifier(CuratorDb curatorDb, Mailer mailer) {
+        this.curatorDb = curatorDb;
+        this.mailer = mailer;
+    }
+
+    public void dispatch(Notification notification) {
+        var tenant = curatorDb.readTenant(notification.source().tenant());
+        tenant.stream().forEach(t -> {
+            if (t instanceof CloudTenant) {
+                var ct = (CloudTenant) t;
+                ct.info().contacts().all().stream()
+                        .filter(c -> c.audiences().contains(TenantContacts.Audience.NOTIFICATIONS))
+                        .collect(Collectors.groupingBy(TenantContacts.Contact::type, Collectors.toList()))
+                        .entrySet()
+                        .forEach(e -> dispatch(notification, e.getKey(), e.getValue()));
+            }
+        });
+    }
+
+    private void dispatch(Notification notification, TenantContacts.Type type, Collection<? extends TenantContacts.Contact> contacts) {
+        switch (type) {
+            case EMAIL:
+                dispatch(notification, contacts.stream().map(c -> (TenantContacts.EmailContact) c).collect(Collectors.toList()));
+                break;
+            default:
+                log.fine("Unknown TenantContact Type: " + type);
```

Comment:
This should throw instead so that sending notifications for unknown contact types doesn't fail silently.

Reply:
I was a bit back and forth on that. Throwing here will also block a notification being persisted to zookeeper. Is that OK? We can alternatively move the dispatch to after the ZK transaction has been committed. But that again can lead to lost mails..

Reply:
Sending can fail for other reasons too, i.e. the mailer service being unavailable. So it's better that sending the notification happens after the write and outside the write lock. Log an error if it fails.

If lost emails becomes a problem we should create a queue for them, and a separate maintainer that consumes it (and retries on failures).